pkg/daemon/server/service/runtime/runtime.go

```diff
@@ -0,0 +1,199 @@
+	r.log.Infof("Starting runtime server...")
+
+	ctx, cancel := context.WithCancel(logging.WithLogger(ctx, r.log))
+
+	go func() {
+		err := r.podTracker.Start(ctx)
+		if err != nil {
+			r.log.Errorw("Failed to start pod tracker for runtime server", zap.Error(err))
+			cancel()
+		}
+	}()
+
+	// start persisting errors into the local cache
+	go r.persistRuntimeErrors(ctx)
+
+	return nil
+}
+
+// persistRuntimeErrors updates the local cache with the runtime errors
+func (r *pipelineRuntimeCache) persistRuntimeErrors(ctx context.Context) {
+	// Define a function to fetch and persist errors
+	fetchAndPersistErrors := func() {
+		var wg sync.WaitGroup
+		for _, vtx := range r.pipeline.Spec.Vertices {
+			for i := range r.podTracker.GetActivePodsCountForVertex(vtx.Name) {
+				wg.Add(1)
+				go r.fetchAndPersistErrorForPod(vtx.Name, i, &wg)
+			}
+		}
+		wg.Wait()
+	}
+
+	// invoke once and then periodically update the cache
+	fetchAndPersistErrors()
+	// Set up a ticker to run periodically
+	ticker := time.NewTicker(runtimeErrorsTimeStep)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			fetchAndPersistErrors()
+		// If the context is done, return.
+		case <-ctx.Done():
+			r.log.Info("Context canceled, stopping PersistRuntimeErrors")
+			return
+		}
+	}
+}
+
+// fetchAndPersistErrorForPod fetches the runtime errors for a pod and persists them in the local cache.
+func (r *pipelineRuntimeCache) fetchAndPersistErrorForPod(vtxName string, podIndex int, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	podName := strings.Join([]string{r.pipeline.Name, vtxName, fmt.Sprintf("%d", podIndex)}, "-")
+	url := fmt.Sprintf("https://%s.%s.%s.svc:%v/%s", podName, r.pipeline.Name+"-"+vtxName+"-headless", r.pipeline.Namespace, v1alpha1.VertexMonitorPort, runtimeErrorsPath)
+
+	res, err := r.httpClient.Get(url)
+	if err != nil {
+		r.log.Errorw("Error reading the runtime errors endpoint: %f", err.Error())
+		return
+	}
+
+	// Read the response body
+	body, err := io.ReadAll(res.Body)
+	_ = res.Body.Close()
+	if err != nil {
+		r.log.Errorf("Error reading response body from %s: %v", url, err)
+		return
+	}
+
+	// Parse the response body into runtime api response
+	var apiResponse ErrorApiResponse
+	if err := json.Unmarshal(body, &apiResponse); err != nil {
+		r.log.Errorf("Error decoding runtime error response from %s: %v", url, err)
+		return
+	}
+
+	if apiResponse.ErrMessage != "" {
+		return
+	}
+
+	cacheKey := PodReplica(fmt.Sprintf("%s-%s-%v", r.pipeline.Name, vtxName, podIndex))
+
+	// Lock the cache before updating
+	r.cacheMutex.Lock()
+	defer r.cacheMutex.Unlock()
+
+	// overwrite the errors
+	r.localCache[cacheKey] = apiResponse.Data
+	r.log.Infof("Persisting error in local cache for: %s", cacheKey)
+}
+
+// GetLocalCache returns the local cache of runtime errors.
+func (r *pipelineRuntimeCache) GetLocalCache() map[PodReplica][]ErrorDetails {
+	r.cacheMutex.RLock()
```

Comment:
What is the point of doing this? When reading items in the return map, the lock has been released.